node/src/effect.rs

```diff
@@ -1749,7 +1735,11 @@ impl<REv> EffectBuilder<REv> {
             // we don't support getting the validators from before the last emergency restart
             return None;
         }
-        self.get_key_block_header_for_era_id_from_storage(era_id)
+        // Era 0 contains no blocks other than the genesis immediate switch block which can be used
+        // to get the validators for era 0.  For any other era `n`, we need the switch block from
+        // era `n-1` to get the validators for `n`.
```

Comment:
I wonder whether it would be better to not sign the genesis block at all.
Signing it this way suggests to the user that it would be safe to request the block with height 0, take the validator set it contains, and then check the corresponding signatures.

Unless you compute the genesis block yourself, you always need a trusted hash, so you never need to check the genesis block's signatures.

Reply:
I see your point, but I'm a bit torn on it.  I mildly prefer the consistency of signing all blocks, and ultimately the genesis block will always be an anomaly in that it doesn't have a parent (we use a sentinel value for the parent hash rather than omitting it via `None` for example)  We're also already signing the genesis block in current and previous versions of the code, and your point applies equally there I think?

I'll wait for feedback from other reviewers before actioning this thanks :)

Reply:
What's the difference between Genesis block (is that a switch block created right after the start? In the past there was no Genesis block) and a switch block right after an emergency upgrade? We should be consistent w.r.t. to security aspect rather than UX in those cases in my opinion.

Reply:
I just find it misleading to take the expected signers out of the block itself. It's fine in this case because we created the block ourselves and the validator set in the genesis block happens to be the actual set of validators for era 0. But what determines that validator set is really the original `accounts.toml` file, not the set specified in the genesis block header. I also think that was how it was originally implemented, when there were no immediate switch blocks. And I think that _if_ we sign the genesis block, we should use the validator set from there, to avoid misleading (even though correct) code.

I agree that the switch block after an emergency restart has the same problem.

In both cases, there are only two ways to trust the block:
* Either you have the `accounts.toml` or the emergency upgrade specification, in which case you might as well compute the block yourself.
* Or you have a trusted hash from after genesis/the upgrade.

I don't see any conceivable case where you could make any use of the signatures.

Anyway, it's not a blocker.